utils/fix_xc7_carry.py

```diff
@@ -0,0 +1,595 @@
+    CARRY4.
+
+    VPR does not currently support either of those options, so for now, if
+    both CO and O are used, the CO output is converted into a LUT equation to
+    recompute the CO output from O, DI and S.  See carry_map.v and
+    clean_carry_map.v for details.
+
+    This script identifies CARRY4 chains in the netlist, identifies if there
+    is output congestion on the O and CO ports, and marks the congestion by
+    changing CARRY_CO_DIRECT (e.g. directly use the CO port) to CARRY_CO_LUT
+    (compute the CO value using a LUT equation).
+
+
+Diagram showing one row of the 7-series CLE, focusing on O/CO congestion.
+
+                                      CLE Row
+
++--------------------------------------------------------------------------+
+|                                                                          |
+|                                                                          |
+|                                               +---+                      |
+|                                               |    +                     |
+|                                               |     +                    |
+|                                     +-------->+ O    +                   |
+|              CO CHAIN               |         |       +                  |
+|                                     |         |       +---------------------> xMUX
+|                 ^                   |   +---->+ CO    +                  |
+|                 |                   |   |     |      +                   |
+|                 |                   |   |     |     +                    |
+|       +---------+----------+        |   |     |    +                     |
+|       |                    |        |   |     +---+                      |
+|       |     CARRY ROW      |        |   |                                |
+|  +--->+ S              O   +--------+   |       xOUTMUX                  |
+|       |                    |        |   |                                |
+|       |                    |        +   |                                |
+|  +--->+ DI             CO  +-------+o+--+                                |
+|       |      CI CHAIN      |        +   |                                |
+|       |                    |        |   |                                |
+|       +---------+----------+        |   |       xFFMUX                   |
+|                 ^                   |   |                                |
+|                 |                   |   |     +---+                      |
+|                 +                   |   |     |    +                     |
+|                                     |   +     |     +    +-----------+   |
+|                                     +--+o+--->+ O    +   |           |   |
+|                                         +     |       +  |    xFF    |   |
+|                                         |     |       +->--D----   Q +------> xQ
```

Comment:
Something looks a little broken here on this line of the diagram? Is it just github?

Reply:
Are you viewing the raw file or the diff?  Also it needs to be in monospace font.